src/clib/pio_spmd.c

```diff
@@ -305,108 +251,97 @@ int pio_swapm(void *sndbuf, int *sndlths, int *sdispls, MPI_Datatype *stypes,
-	{
-            if (stypes[i] == MPI_DATATYPE_NULL)
-                stypes[i] = MPI_CHAR;
-            if (rtypes[i] == MPI_DATATYPE_NULL)
-                rtypes[i] = MPI_CHAR;
+        /* For some reason OPEN_MPI has problems with MPI_CHAR ??? */
+        for (int i = 0; i < ntasks; i++)
+        {
+            if (sendtypes[i] == MPI_DATATYPE_NULL)
+                sendtypes[i] = MPI_CHAR;
+            if (recvtypes[i] == MPI_DATATYPE_NULL)
+                recvtypes[i] = MPI_CHAR;
         }
 #endif
-        CheckMPIReturn(MPI_Alltoallw(sndbuf, sndlths, sdispls, stypes, rcvbuf,
-				     rcvlths, rdispls, rtypes, comm),__FILE__,__LINE__);
+
+        /* Call the MPI alltoall without flow control. */
+        CheckMPIReturn(MPI_Alltoallw(sendbuf, sendcounts, sdispls, sendtypes, recvbuf,
+                                     recvcounts, rdispls, recvtypes, comm), __FILE__, __LINE__);
 
 #ifdef OPEN_MPI
-        for (int i = 0; i < nprocs; i++)
-	{
-            if (stypes[i] == MPI_CHAR)
-                stypes[i] = MPI_DATATYPE_NULL;
-            if (rtypes[i] == MPI_CHAR)
-                rtypes[i] = MPI_DATATYPE_NULL;
+        /* For some reason OPEN_MPI has problems with MPI_CHAR ??? */
+        for (int i = 0; i < ntasks; i++)
+        {
+            if (sendtypes[i] == MPI_CHAR)
+                sendtypes[i] = MPI_DATATYPE_NULL;
+            if (recvtypes[i] == MPI_CHAR)
+                recvtypes[i] = MPI_DATATYPE_NULL;
         }
 #endif
         return PIO_NOERR;
     }
 
-    int tag;
-    int offset_t;
-    int ierr;
-    MPI_Status status;
-    int steps;
-    int istep;
-    int rstep;
-    int p;
-    int maxreq;
-    int maxreqh;
-    int hs;
-    int cnt;
-    void *ptr;
-    MPI_Request rcvids[nprocs];
+    /* ??? */
+    offset_t = ntasks;
 
-    offset_t = nprocs;
-    // send to self
-    if (sndlths[mytask] > 0)
+    /* Send to self. */
+    if (sendcounts[my_rank] > 0)
     {
         void *sptr, *rptr;
         int extent, lb;
-        tag = mytask + offset_t;
-        sptr = (void *)((char *) sndbuf + sdispls[mytask]);
-        rptr = (void *)((char *) rcvbuf + rdispls[mytask]);
+        tag = my_rank + offset_t;
+        sptr = (char *)sendbuf + sdispls[my_rank];
+        rptr = (char *)recvbuf + rdispls[my_rank];
 
         /*
-          MPI_Type_get_extent(stypes[mytask], &lb, &extent);
+          MPI_Type_get_extent(sendtypes[my_rank], &lb, &extent);
           printf("%s %d %d %d\n",__FILE__,__LINE__,extent, lb);
-          MPI_Type_get_extent(rtypes[mytask], &lb, &extent);
+          MPI_Type_get_extent(recvtypes[my_rank], &lb, &extent);
           printf("%s %d %d %d\n",__FILE__,__LINE__,extent, lb);
         */
-#ifdef ONEWAY
-        CheckMPIReturn(MPI_Sendrecv(sptr, sndlths[mytask],stypes[mytask],
-                                    mytask, tag, rptr, rcvlths[mytask], rtypes[mytask],
-                                    mytask, tag, comm, &status),__FILE__,__LINE__);
+#ifdef ONEWAY /* ??? */
```

Comment:
ONEWAY or another - this is a bad name for this macro which if true will post mpi_sendrecv comms instead of irecv/send.